master/pkg/checkpoints/gcs/gcs_checkpoints.go

```diff
@@ -0,0 +1,100 @@
+package gcs
+
+import (
+	"context"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"github.com/docker/go-units"
+	"google.golang.org/api/iterator"
+
+	"github.com/determined-ai/determined/master/pkg/checkpoints/archive"
+)
+
+// GCSDownloader implements downloading a checkpoint from GCS
+// and sends it to the client in an archive file.
+type GCSDownloader struct {
+	aw     archive.ArchiveWriter
+	bucket string
+	prefix string
+	buffer []byte
+}
+
+// DefaultDownloadPartSize is the default part size for downloading files from GCS.
+// This is the same as the default part size for S3.
+const DefaultDownloadPartSize = units.MiB * 5
+
+func (d *GCSDownloader) fileDownload(
+	ctx context.Context,
+	o *storage.ObjectAttrs,
+	r *storage.Reader) error {
+	defer func() {
+		_ = r.Close()
+	}()
+	if err := d.aw.WriteHeader(strings.TrimPrefix(o.Name, d.prefix), o.Size); err != nil {
+		return err
+	}
+	remaining := o.Size
+	for {
+		if remaining <= 0 {
+			break
+		}
+		sizeRead, err := r.Read(d.buffer)
+		if err != nil {
+			return err
+		}
+		if _, err := d.aw.Write(d.buffer[:sizeRead]); err != nil {
+			return err
+		}
+		remaining -= int64(sizeRead)
+	}
+	return nil
+}
+
+// Download downloads the checkpoint.
+func (d *GCSDownloader) Download(ctx context.Context) error {
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return err
```

Comment:
Should we be wrapping these errors, so if a user shows us error logs, that we can properly trace where the error came from?